flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java

```diff
@@ -0,0 +1,135 @@
+    private final KubernetesClient kubernetesClient;
+    private final FlinkService flinkService;
+
+    public FlinkSessionJobReconciler(
+            KubernetesClient kubernetesClient,
+            FlinkService flinkService,
+            FlinkOperatorConfiguration operatorConfiguration) {
+        this.kubernetesClient = kubernetesClient;
+        this.flinkService = flinkService;
+        this.operatorConfiguration = operatorConfiguration;
+    }
+
+    @Override
+    public void reconcile(
+            FlinkSessionJob flinkSessionJob, Context context, Configuration defaultConfig)
+            throws Exception {
+
+        FlinkSessionJobSpec lastReconciledSpec =
+                flinkSessionJob.getStatus().getReconciliationStatus().getLastReconciledSpec();
+
+        if (lastReconciledSpec == null) {
+            deployFlinkJob(flinkSessionJob, defaultConfig);
+            return;
+        }
+
+        boolean specChanged = !flinkSessionJob.getSpec().equals(lastReconciledSpec);
+
+        if (specChanged) {
+            // TODO reconcile other spec change.
+            LOG.info("Other spec change have not supported");
+        }
+    }
+
+    @Override
+    public DeleteControl cleanup(FlinkSessionJob sessionJob, Configuration defaultConfig) {
+        Optional<FlinkDeployment> flinkDepOptional = getFlinkDeployment(sessionJob);
+
+        if (flinkDepOptional.isPresent()) {
+            Configuration effectiveConfig =
+                    FlinkUtils.getEffectiveConfig(flinkDepOptional.get(), defaultConfig);
+            String jobID = sessionJob.getStatus().getJobStatus().getJobId();
+            if (jobID != null) {
+                try {
+                    flinkService.cancelSessionJob(JobID.fromHexString(jobID), effectiveConfig);
+                } catch (Exception e) {
+                    LOG.error("Failed to cancel job.", e);
+                }
+            }
+        } else {
+            LOG.info("Session cluster deployment not available");
+        }
+        return DeleteControl.defaultDelete();
+    }
+
+    private void deployFlinkJob(FlinkSessionJob sessionJob, Configuration defaultConfig)
+            throws Exception {
+        Optional<FlinkDeployment> flinkDepOptional = getFlinkDeployment(sessionJob);
+        if (flinkDepOptional.isPresent()
+                && (flinkDepOptional.get().getStatus().getJobManagerDeploymentStatus()
+                        == JobManagerDeploymentStatus.READY)) {
+            Configuration effectiveConfig =
+                    FlinkUtils.getEffectiveConfig(flinkDepOptional.get(), defaultConfig);
+            flinkService.submitJobToSessionCluster(sessionJob, effectiveConfig);
+            ReconciliationUtils.updateForSpecReconciliationSuccess(sessionJob);
+        } else {
+            LOG.info("Session cluster deployment is not in READY state");
+        }
+    }
+
+    private Optional<FlinkDeployment> getFlinkDeployment(FlinkSessionJob sessionJob) {
```

Comment:
I have tried this way a bit, but not quite work out, I will look into this more deeply

Reply:
Technically the  event source from the `FlinkDeployment` will have to get which `FlinkSessionJob` to trigger, this will also have to fetch from the apiserver or get from the informer cache. (Because we can't track the target `FlinkSessionJob` based on the `FlinkDeployment` CR ) 
In the case of `FlinkDeploumenController`, the change of `Deployment` can be track to the target `FlinkDeployment` by the deployment metadata.
On the other hand, whether we need to register new source for `FlinkDeployment` should depend on whether we have to react to the change to the `FlinkDeployment`. As we need the upgrade and delete of the session only work with there is no active session job, So I think we do not have the requirements to register a event source for session `FlinkDeployment`. 
But we truly need to get the target `FlinkDeployment` in a lightweight way, I prefer to use an informer to cache the `FlinkDeployment` for it .

What do you think of this ?

Reply:
What's the difference between register an `InformerEventSource` with manual created informer? I just think register a new event source and `getSecondaryResource()` is easier to understand.

Reply:
Register an InformerEventSource firstly need to create an informer and then tell the SDK how to mapping the secondaryResource to the custom resource. By this, when a secondaryResource changed, it can trigger an event for the custom resource (session job)

Reply:
I rethink about this, I think register a new event source for the flinkdep seems better, by this, we can accelerate the reaction time when a  session job cluster ready for submit, I will submit a commit for it.

Reply:
Yes. What I mean is using `InformerEventSource` will not take any additional burden to K8s APIServer, compared with naked informer.